api/types.go

```diff
@@ -35,6 +35,22 @@ func NewTagSet() TagSet {
 	return make(map[string]string)
 }
 
+func (left TagSet) Equals(right TagSet) bool {
+	for k := range left {
+		_, ok := right[k]
+		if !ok || left[k] != right[k] {
+			return false
+		}
+	}
+	for k := range right {
```

Comment:
This part can probably just compare `len(left) == len(right)` and we'd have the same behavior.